include/pybind11/detail/common.h

```diff
@@ -280,6 +291,17 @@ extern "C" {
             });
         }
 \endrst */
+#if defined(PYBIND11_NOEXCEPTIONS)
+#define PYBIND11_MODULE(name, variable)                                        \
+    static void PYBIND11_CONCAT(pybind11_init_, name)(pybind11::module &);     \
+    PYBIND11_PLUGIN_IMPL(name) {                                               \
+        PYBIND11_CHECK_PYTHON_VERSION                                          \
+        auto m = pybind11::module(PYBIND11_TOSTRING(name));                    \
```

Comment:
So what happens now if constructing an extension module throws?

Imagine a line inside a user's `PYBIND11_MODULE` that says `py::module foo = py::import("throwy_module");` and actually having a exception thrown from the `throwy_module`.

Reply:
If you've set the flag PYBIND11_NOEXCEPTIONS and you run without enable exceptions I think it will fail to compile.

Reply:
I was talking about raising an exception from pure python, which can't be detected at compile time. I gave it a test and here are the two similar cases that I've tried:

```c++
// foo.cpp
#include <pybind11/pybind11.h>
PYBIND11_MODULE(foo, m) {
	pybind11::module x = pybind11::module::import("boom");
}
```
If there is no `boom.py` in python's path, the following happens:

```
Python 3.7.3 (default, Mar 27 2019, 19:19:44)
[GCC 8.2.1 20181127] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> try:
...     import foo
... except Exception:
...
KeyboardInterrupt
>>> try:
...     import foo
... except Exception as e:
...     print(e)
...
terminate called after throwing an instance of 'pybind11::error_already_set'
  what():  ModuleNotFoundError: No module named 'boom'
zsh: abort      python
```

And if we add a `boom.py` that looks like this:

```python
raise RuntimeError
```

the error becomes more verbose:

```
Python 3.7.3 (default, Mar 27 2019, 19:19:44)
[GCC 8.2.1 20181127] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import boom
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/bstaletic/Temp/pybind11/boom.py", line 1, in <module>
    raise RuntimeError
RuntimeError
>>> try:
...     import foo
... except Exception as e:
...     print(e)
...
terminate called after throwing an instance of 'pybind11::error_already_set'
  what():  RuntimeError:

At:
  /home/bstaletic/Temp/pybind11/boom.py(1): <module>
  <frozen importlib._bootstrap>(219): _call_with_frames_removed
  <frozen importlib._bootstrap_external>(728): exec_module
  <frozen importlib._bootstrap>(677): _load_unlocked
  <frozen importlib._bootstrap>(967): _find_and_load_unlocked
  <frozen importlib._bootstrap>(983): _find_and_load
  <frozen importlib._bootstrap>(219): _call_with_frames_removed
  <frozen importlib._bootstrap_external>(1043): create_module
  <frozen importlib._bootstrap>(583): module_from_spec
  <frozen importlib._bootstrap>(670): _load_unlocked
  <frozen importlib._bootstrap>(967): _find_and_load_unlocked
  <frozen importlib._bootstrap>(983): _find_and_load
  <stdin>(2): <module>

zsh: abort      python
```

Either way, it quickly calls `std::terminate()`, which is completely fine, but let's mention it in the changelog too.

Reply:
Ah whoops, I totally misread that. Okay I'll update the changelog. Thanks @bstaletic